dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/TracedEncoding.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.conjure.java.dialogue.serde;
+
+import com.palantir.dialogue.TypeMarker;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.tracing.CloseableTracer;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.lang.reflect.Type;
+
+/**
+ * Wrapper around an {@link Encoding} which adds tracing spans around serialization and deserialization operations
+ * including I/O.
+ */
+final class TracedEncoding implements Encoding {
+
+    private final Encoding delegate;
+
+    TracedEncoding(Encoding delegate) {
+        this.delegate = Preconditions.checkNotNull(delegate, "Encoding is required");
+    }
+
+    @Override
+    public <T> Serializer<T> serializer(TypeMarker<T> type) {
+        String operation = "Dialogue: serialize " + toString(type) + " to " + getContentType();
+        return new TracedSerializer<>(delegate.serializer(type), operation);
+    }
+
+    @Override
+    public <T> Deserializer<T> deserializer(TypeMarker<T> type) {
+        String operation = "Dialogue: deserialize " + toString(type) + " from " + getContentType();
+        return new TracedDeserializer<>(delegate.deserializer(type), operation);
+    }
+
+    @Override
+    public String getContentType() {
+        return delegate.getContentType();
+    }
+
+    @Override
+    public boolean supportsContentType(String contentType) {
+        return delegate.supportsContentType(contentType);
+    }
+
+    @Override
+    public String toString() {
+        return "TracedEncoding{delegate=" + delegate + '}';
```

Comment:
i've actually started omitting the `delegate=` when there's only one field, especially because we compose so much stuff you end up with a 
`Foo{delegate=Bar{delegate=Baz{delegate=`
vs 
`Foo{Bar{Baz{`

Reply:
that's much  more concise, updated this along with the lazily initialized encoding.